### ISSUE TITLE
LEfSe - subgroup check did not remove non-significant taxa

### ISSUE THREAD

**Issue by AutumnPears** [ChiLiubio/microeco#399]
Hello,

I have a question about performing LEfSe with trans_diff$new. My microeco version is 1.9.0. My dataset has two treatments (two groups), and each has two subgroups (high frequency and low frequency). My code is below:

meco.NNB <- phyloseq2meco(ps.NNB)    #import data from phyloseq
lefN <- trans_diff$new( dataset = meco.NNB,
                        method = "lefse",
                        group = "NBtype",
                        lefse_subgroup = "Freq",
                        lefse_min_subsam = 4,
                        p_adjust_method = "none",
                        alpha = 0.05)

Here is the log while the code was running:
No taxa_abund list found. Calculate relative abundance with cal_abund function ...
The row number of tax_table is not equal to that of otu_table ...
Automatically applying tidy_dataset() function to trim the data ...
microtable-class object:
sample_table have 24 rows and 5 columns
otu_table have 1906 rows and 24 columns
tax_table have 1906 rows and 6 columns
The result is stored in object$taxa_abund ...
1294 input features ...
854 features are remained after removing unknown features ...
Start Kruskal-Wallis rank sum test for NBtype ...
87 taxa found significant ...
After P value adjustment, 87 taxa found significant ...
Start lefse subgroup biomarkers check for Freq ...
Remove 0 biomarkers after subgroup check ...
Minimum LDA score: 1.83936511685227 maximum LDA score: 4.35023740483947
Taxa abundance table is stored in object$res_abund ...
lefse analysis result is stored in object$res_diff ...
There were 50 or more warnings (use warnings() to see the first 50)

I am surprised that no taxa was removed after subgroup check, so I checked some of the identified taxa using the abundance table ($res_abund) and found that some taxa from $res_diff did not test significant by the Wilcox rank sum test in subgroups (some had p>0.05 in one of the two subgroups, some had p>0.05 in both subgroups). The way I understand it, after the initial KW test, Wilcoxon tests will be carried out for each subgroup (so comparing subgroup 1 between 2 treatment, then comparing subgroup 2 between two treatments). Only taxa with p < 0.05 in all three tests would be retained in $res_diff. Is that correct?

I tried the above code again with alpha = 0.5, and the log says that 557 taxa found significant ...Remove 56 biomarkers after subgroup check. So the Wilcox tests are being done, but removal criteria seems off. 

Could the problem be with my sample size? In my treatment 2, each subgroup only has 4 samples (in treatment 1, each subgroup has 8 samples. Total: 24 samples). 

I really like the cladograms plotted with this package, so I hope you could help me solve this issue.

**Comment by ChiLiubio** [comment]
Hi. The goal of subgroup checking in lefse is to remove the features that donot distribute homogeneously across subgroups in each treatment. So it is not "so comparing subgroup 1 between 2 treatment". I think your code is fine to run.

Best,
Chi

**Comment by NNTLe** [comment]
Hello,

Thank you so much for your prompt reply. I learned to use lefse recently so my understanding may not be accurate. I read the paper "Metagenomic biomarker discovery and explanation" by Segata et al. (https://genomebiology.biomedcentral.com/articles/10.1186/gb-2011-12-6-r60#auth-Nicola-Segata-Aff1), and it said:

"The pairwise Wilcoxon test is applied to retained features belonging to subclasses of different classes. For each feature, the pairwise Wilcoxon test is not satisfied if at least one comparison between subclasses has a P-value higher than the chosen α or if the sign of variation is not equal among all comparisons. For example, if a feature appears in samples from two classes with three subclasses each, all nine comparisons between subclasses in different classes must violate the null hypothesis, and all signs of the differences between medians must be consistent. The features that pass the pairwise Wilcoxon test are considered successful biomarkers."

From this description, I though that after the KW test comparing treatment 1 to treatment 2, you would do Wilcoxon tests comparing treatment 1-subgroup 1 to treatment 2-subgroup 1, then treatment 1-subgroup 2 to treatment 2-subgroup 2, and so on, and keep only the features with p<0.05 in all tests (this is the less strict version comparing subgroups of the same name, compared to the text above that suggests Wilcoxon tests between all pairs of subgroups). 

What do you think about this? Does the lefse method in the microeco package use a different protocol than the Segata paper?

**Comment by ChiLiubio** [comment]
Yes. You are right. So sorry. I forgot the details of this method. I checked the whole code in the microeco package again. It is totally same with the paper described. Beside "treatment 1-subgroup 1 to treatment 2-subgroup 1, then treatment 1-subgroup 2 to treatment 2-subgroup 2", there will be also something like "treatment 1-subgroup 1 to treatment 2-subgroup 2".
Back to your initial issue "then comparing subgroup 2 between two treatments). Only taxa with p < 0.05 in all three tests would be retained in $res_diff. Is that correct?", it is better to understand if we say the subgroup test is comparing medians in two groups and p values. If a different case is found (e.g., median of treatment 1-subgroup 1 > treatment 2-subgroup 1, but treatment 1 < treatment 2 for all), the function then check the p value. If the p value is significant, the feature will be removed.
If the data distribute homogeneously across subgroups, there will be actually very little taxa that should be removed.

**Comment by NNTLe** [comment]
Hello, thank you for your explanation! The results I got from microeco made sense now. 

However, I can't help but feel that the method you explained is more lenient than the Segata paper. As quoted in my previous reply, the Segata paper required both a significant Wilcoxon p values between subgroups (treatment 1-subgroup 1 to treatment 2-subgroup 1, etc) and the trend of the medians to be the same as the KW. In your explanation, it sounds like if the trend of the medians agree with KW (which is the easier requirement to meet), you would not check the Wilcoxon p values. Therefore, I imagine that some of the 87 taxa I found with the microeco would not be detected as biomarkers based on the Segata paper, since they have the same trend of the medians between subgroups, but the Wilcoxon p values between subgroups were not significant. Would you agree with this?

Also, when you said "If the p value is significant, the feature will be removed", you are comparing two subgroups within the same treatment, right (Treatment 1-subgroup 1 to Treatment1-subgroup 2)? So removing features with significant p values means that the features that are not distributed homogeneously between subgroups in the same treatment are removed. This is not the same p value mentioned in the Segata paper, which requires "all nine comparisons between subclasses in different classes must violate the null hypothesis". I think the Segata paper means significant p values between Treatment 1-subgroup 1 to Treatment 2-subgroup 1 etc. are required.

**Comment by ChiLiubio** [comment]
Hi. I originally planned to use the online LEfSe on Galaxy to test and compare the results of the subclass, but I was not successful. When it becomes feasible, I will compare it again. I suspect that the method in microeco and the Segata paper is similar because the requirement for all subclass combinations to be significant is too stringent, and the sample size will greatly affect the results. This can also be seen from the descriptions in the literature.
![图片](https://github.com/user-attachments/assets/fc6e71a7-13f5-4cdb-baa2-ecbf997c5176)

The comparison method is exactly the same as in the Segata paper. What I mentioned about "not being distributed homogeneously" refers to the purpose of this detection, not the process. All comparisons are made between different classes (Treatment 1 sth -Treatment 2 sth).

**Comment by NNTLe** [comment]
I have a different understanding of the screenshoted paragraph. It says the pairwise Wilcoxon test is not satisfied if at least one comparison has P value higher than the significance level (usually 0.05), or the sign of the variation is not the same among all comparisons. The Wilcoxon test is not satisfied means the feature is rejected, no? If either of the above happens, the feature is rejected. 

Let's say feature A is higher in Treatment 2 compared to Treatment 1 with KW p < 0.05. Then, according to the screenshot, if A has p > 0.05 (not significant) in at least one of the pairwise Wilcoxon tests, A would be rejected (i.e. Wilcoxon test not satisfied). If all Wilcoxon p values are significant but A is higher in Treatment 1 than in Treatment 2 in one subgroup comparison, A is also rejected. The screenshot mentions p value checking before sign checking. It does not say p value checking is unnecessary if the signs agree. In addition, Segata et al reject A if p > 0.05 in subgroup test, while you said you rejected A if p < 0.05 (and the sign is wrong), so it's not the same. 

Have you tried checking with the MicrobiomeMarker package? I only got 32 significant features with the same dataset using that package, and these include unclassified taxa.

**Comment by ChiLiubio** [comment]
Oh. Thank you for the reminder. The official version should be as you described. However, it might be as I initially thought, that if the checking for subclasses is set too strictly, it will filter out too many features, and the sample size in subclass will greatly affect these tests, with few subclasses having a large number of samples. Therefore, I believe it is sufficient to consider only the cases with significant inconsistencies (significnat p with different sign direction), that is, when the data has a relatively extreme distribution imbalance. For those cases with some significant and some non-significant features in subclass test, it usually indicates that the two groups do not differ significantly overall, and these features, even if retained, will have a relatively low score in subsequent LDA analysis. It is also notable that another difference between the method in microeco and the Python version in the article is that the p-values here are adjusted by default.
I will consider to add a parameter to adjust the subclass test way (such as strict mode like the paper shows or loose mode like the current set).

**Comment by NNTLe** [comment]
I see. Thank you for the clarification.
